### PR TITLE
core/sched: add runq_callback hook and runq inspection functions

### PR DESCRIPTION
**Pull request by kfessel**
### Contribution description

This adds a runqueue callback to core/sched that is called on changes to runqueue if they are active or their last thread is removed. (callbacks are issued when: a runqueue gets active, the active runqueue grows or a runqueue is emptied)

This also adds functions to gather informations about the length of runqueues. (empty, one thread, more than one thread)

Both features together may be used to implement schedulers that are able to adapt to demand (active if required, inactive if not)

### Testing procedure

none yet but testing #16126

### Issues/PRs references

#16311 added the required interface to change threads by advancing the runqueue

this is derived from #16126 which also adds a matching round robin scheduler

### CODE REVIEW

File: core/include/sched.h [hunk 1]
```diff
@@ -298,6 +298,62 @@ static inline void sched_runq_advance(uint8_t prio)
...
+/**
+ * @brief   Tell if the number of threads in a runqueue is 1
+ *
+ * @param[in]   prio      The priority of the runqueue to get information of
+ * @return      Truth value for that information
+ * @warning     This API is not intended for out of tree users.
+ */
+static inline int sched_runq_is_one(uint8_t prio)
+{
+    return clist_exactly_one(&sched_runqueues[prio]);
```
Comment (maribu) [inline core/include/sched.h:342]: IMO using consistent postfixes (e.g. `sched_runq_exactly_one()`) makes life a bit easier.

File: core/include/sched.h [hunk 2]
```diff
@@ -298,6 +298,62 @@ static inline void sched_runq_advance(uint8_t prio)
...
+/**
+ * @brief   Tell if the number of threads in a runqueue greater than 1
+ *
+ * @param[in]   prio      The priority of the runqueue to get information of
+ * @return      Truth value for that information
+ * @warning     This API is not intended for out of tree users.
+ */
+static inline int sched_runq_greater_one(uint8_t prio)
+{
+    return clist_more_than_one(&sched_runqueues[prio]);
```
Comment (maribu) [inline core/include/sched.h:354]: Same

File: core/sched.c [hunk 3]
```diff
@@ -156,6 +156,10 @@ thread_t *__attribute__((used)) sched_run(void)
     thread_t *next_thread = container_of(sched_runqueues[nextrq].next->next,
                                          thread_t, rq_entry);
 
+#if (IS_USED(MODULE_SCHED_RUNQ_CALLBACK))
```
Comment (maribu) [inline core/sched.c:159]: Would you mind to sneak in the style fix suggested by the CI for lines 78/79? (But as separate commit, to ease potential reverts.)

**Review by maribu — COMMENTED** [review]
Looks good to me.

I wonder if this could be more generalized to e.g. also aid a potential statistics module. But since the API is marked as only suitable for internal use, it might indeed be better to wait for such a user to appear (if ever) and generalize the callbacks then. At least I tent to not get APIs right without having all code using it in every intended way in place.

**Comment by kfessel** [comment]
on rebase (squash) i will move the atm last commit to the begin and squash the remaining

**Review by maribu — APPROVED** [review]
@kaspar030: Care to have a look?

**Review by benpicco — APPROVED** [review]
The code looks good and changes do not affect the code by default.
Unless @kaspar030 raises objections, I see no reason to go ahead with this.

**Comment by kfessel** [comment]
I will merge this tomorrow  (7 days without the possibly to come objection)

**Review by kaspar030 — APPROVED** [review]
ACK.

looks fine to me. and as is it is optional.

**Comment by kaspar030** [comment]
Thanks for waiting!

**Pull request merged**